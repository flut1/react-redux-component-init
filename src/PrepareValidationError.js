import ExtendableError from './utils/ExtendableError';

class PrepareValidationError extends ExtendableError {}

export default PrepareValidationError;
