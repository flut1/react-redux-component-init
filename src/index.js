export { default as withInitAction } from './withInitAction';
export { default as initReducer } from './reducer';

export { default as initComponent } from './actions/initComponent';
export { default as prepareComponent } from './actions/prepareComponent';
export { default as prepareComponents } from './actions/prepareComponents';
export { default as setInitMode } from './actions/setInitMode';

export { default as PrepareValidationError } from './PrepareValidationError';
export * from './initMode';
export * from './initSelfMode';
export * as actionTypes from './actions/actionTypes';
