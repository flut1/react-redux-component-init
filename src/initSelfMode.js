/**
 * Possible values for the `selfInit` option of `withInitAction()`
 */

/**
 * The component will render immediately, even if `initAction` is still pending.
 * It is recommended to use this option and render a loading indicator or placeholder content
 * until `initAction` is resolved. This will give the user immediate feedback that something is
 * being loaded. While the `initAction` is pending, an `isInitializing` prop will be passed to
 * the component.
 * @type {string}
 */
export const INIT_SELF_ASYNC = 'ASYNC';

/**
 * This will cause the component not to be mounted until the first initialization has completed.
 * The component will remain mounted during any reinitialization.
 * @type {string}
 */
export const INIT_SELF_BLOCKING = 'BLOCKING';

/**
 * The same behavior as `INIT_SELF_BLOCKING` but it will also unmount the component during
 * reinitialization.
 * @type {string}
 */
export const INIT_SELF_UNMOUNT = 'UNMOUNT';

/**
 * Only execute initAction on the server (initMode == MODE_PREPARE). Initialization will
 * be skipped on the client.
 * @type {string}
 */
export const INIT_SELF_NEVER = 'NEVER';
