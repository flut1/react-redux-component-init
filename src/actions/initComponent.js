import propNameValuesToObject from '../utils/propNameValuesToObject';
import { MODE_INIT_SELF, MODE_PREPARE } from '../initMode';
import { INIT_SELF_NEVER } from '../initSelfMode';

import { INIT_COMPONENT } from './actionTypes';

/**
 * Action that initializes a single Component. Can be used both during a component prepare
 * and a component self-init.
 *
 * If the current `initMode` is `MODE_PREPARE` and the component has not yet initialized
 * when this action is dispatched, an error will be thrown.
 *
 * **There is probably no need to directly use this action outside of this module.** This
 * action is automatically dispatched by the `withInitAction()` HoC and the `prepareComponent()`
 * thunk.
 *
 * @param {react.Component} Component The React component to initialize. Passing a component that
 * was not wrapped with `withInitAction()` will result in an error.
 * @param {Array} initValues An array of values that correspond with the array of `initProps` for
 * this component. These values are either passed through `prepareComponent()` or extracted from
 * the `props` when a component mounts or updates.
 * @param {string} prepareKey A unique identifier for the component and it's `initValues`. This
 * is generated by the `getPrepareKey()` util.
 * @param {object} [options] Additional options
 * @param {boolean} [options.isPrepare = false] True if a component is being prepared, false
 * if a component is being mounted or updated.
 * @returns {function} A thunk function that should be passed directly to the Redux `dispatch`
 * function.
 */
export default (
  Component,
  initValues,
  prepareKey,
  { isPrepare = false } = {},
) => (dispatch, getState) => {
  if (!Component.initConfig) {
    throw new Error('No init config found on Component passed to initComponent');
  }

  const {
    componentId,
    initProps,
    initAction,
    options: { onError, getInitState, initSelf, allowLazy },
  } = Component.initConfig;

  const initState = getInitState(getState());
  if (!initState) {
    throw new ReferenceError('Could not find init state. Did you attach the init reducer?');
  }
  const { mode, prepared } = initState;

  let shouldInitialize = (
    ((mode === MODE_INIT_SELF) && (initSelf !== INIT_SELF_NEVER)) ||
    (isPrepare && (typeof prepared[prepareKey] === 'undefined'))
  );
  if ((mode === MODE_PREPARE) && !isPrepare && allowLazy && !prepared[prepareKey]) {
    shouldInitialize = true;
  }

  if (shouldInitialize) {
    const initPropsObj = propNameValuesToObject(initProps, initValues);

    dispatch({
      type: INIT_COMPONENT,
      payload: {
        complete: false,
        isPrepare,
        prepareKey,
      },
    });

    return Promise.resolve()
      .then(() => {
        const initActionReturn = initAction(initPropsObj, dispatch, getState);

        if (typeof initActionReturn.then !== 'function') {
          const error = new Error(`Expected initAction to return a Promise. Returned an ${typeof initActionReturn} instead. Check the initAction for "${componentId}"`);
          error.isInvalidReturnError = true;
          throw error;
        }

        return initActionReturn;
      })
      .catch((e) => {
        if (onError && !e.isInvalidReturnError) {
          onError(e);
        } else {
          throw e;
        }
      })
      .then((result) => {
        dispatch({
          type: INIT_COMPONENT,
          payload: {
            complete: true,
            isPrepare,
            prepareKey,
          },
        });

        return result;
      });
  } else if (mode === MODE_PREPARE && !isPrepare && !allowLazy) {
    if (typeof prepared[prepareKey] === 'undefined') {
      const initPropsObj = propNameValuesToObject(initProps, initValues);
      throw new Error(`Expected component "${componentId}" to be prepared but prepareComponent has not been called with props: \n${JSON.stringify(initPropsObj)}`);
    } else if (prepared[prepareKey] === false) {
      throw new Error(`Expected component "${componentId}" to be prepared but preparation is still pending`);
    }
  }

  return Promise.resolve();
};
