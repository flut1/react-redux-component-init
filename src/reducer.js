import { createSelector } from 'reselect';

import createShallowArrayCompareSelector from './utils/createShallowArrayCompareSelector';
import { INIT_COMPONENT, SET_INIT_MODE } from './actions/actionTypes';
import extractValuesForProps from './utils/extractValuesForProps';
import { MODE_PREPARE } from './initMode';

/**
 * Reducer function that manages the state for `react-redux-component-init`. This reducer should
 * be included using Redux's `combineReducers()` under the `init` key in the root of
 * the store state.
 */
export default (state = {
  mode: MODE_PREPARE,
  prepared: {},
  selfInit: {},
}, action) => {
  switch (action.type) {
    case SET_INIT_MODE:
      return {
        ...state,
        mode: action.payload,
      };
    case INIT_COMPONENT:
      if (action.payload.isPrepare) {
        return {
          ...state,
          prepared: {
            ...state.prepared,
            [action.payload.prepareKey]: action.payload.complete,
          },
        };
      }
      return {
        ...state,
        selfInit: {
          ...state.selfInit,
          [action.payload.prepareKey]: action.payload.complete,
        },
      };
    default:
      return state;
  }
};

const createComponentInitValuesSelector = ({
  componentId, initProps, options: { getPrepareKey },
}) =>
  createShallowArrayCompareSelector(
    (state, props) => extractValuesForProps(props, initProps),
    initValues => ({
      prepareKey: getPrepareKey(componentId, initValues),
      initValues,
    }),
  );

export const createComponentInitStateSelector = initConfig => createSelector(
  createComponentInitValuesSelector(initConfig),
  state => state.selfInit,
  state => state.prepared,
  ({ prepareKey, initValues }, selfInit, prepared) => ({
    prepareKey,
    initValues,
    selfInitState: selfInit[prepareKey],
    isPrepared: !!prepared[prepareKey],
  }),
);
