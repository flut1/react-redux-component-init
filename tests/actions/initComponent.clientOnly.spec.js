/* global expect, describe, it, jest */
import configureMockStore from 'redux-mock-store';
import thunk from 'redux-thunk';

import initComponent from '../../src/actions/initComponent';
import withInitAction, { clearComponentIds } from '../../src/withInitAction';
import { MODE_INIT_SELF, MODE_PREPARE } from '../../src/initMode';
import SimpleInitTestComponent from '../fixtures/SimpleInitTestComponent';

const mockStore = configureMockStore([thunk]);

describe('initComponent with clientOnly action', () => {
  describe('when initMode === MODE_PREPARE', () => {
    describe("with { caller: 'willMount' } option given", () => {
      clearComponentIds();
      const testAction = jest.fn(() => Promise.resolve());
      const store = mockStore({ init: { mode: MODE_PREPARE, prepared: {} } });
      const MockWithInit = withInitAction({ clientOnly: testAction })(SimpleInitTestComponent);

      store.dispatch(
        initComponent(MockWithInit, [], 'SimpleInitTestComponent[]', { caller: 'willMount' }),
      );

      it('does not call the action', () =>
        new Promise(resolve => {
          process.nextTick(() => {
            expect(testAction).not.toHaveBeenCalled();
            resolve();
          });
        }));

      it('does not dispatch INIT_COMPONENT', () => {
        const actions = store.getActions();
        return expect(actions).toEqual([]);
      });
    });
    describe("with { caller: 'didMount' } option given", () => {
      clearComponentIds();
      const testAction = jest.fn(() => Promise.resolve());
      const store = mockStore({ init: { mode: MODE_PREPARE, prepared: {} } });
      const MockWithInit = withInitAction({ clientOnly: testAction })(SimpleInitTestComponent);

      store.dispatch(
        initComponent(MockWithInit, [], 'SimpleInitTestComponent[]', { caller: 'didMount' }),
      );

      it('does not call the action', () =>
        new Promise(resolve => {
          process.nextTick(() => {
            expect(testAction).not.toHaveBeenCalled();
            resolve();
          });
        }));

      it('does not dispatch INIT_COMPONENT', () => {
        const actions = store.getActions();
        return expect(actions).toEqual([]);
      });
    });
  });

  describe('when initMode === MODE_INIT_SELF', () => {
    describe("with { caller: 'willMount' } option given", () => {
      clearComponentIds();
      const testAction = jest.fn(() => Promise.resolve());
      const store = mockStore({ init: { mode: MODE_INIT_SELF, prepared: {} } });
      const MockWithInit = withInitAction({ clientOnly: testAction })(SimpleInitTestComponent);

      store.dispatch(
        initComponent(MockWithInit, [], 'SimpleInitTestComponent[]', { caller: 'willMount' }),
      );

      it('does not call the action', () =>
        new Promise(resolve => {
          process.nextTick(() => {
            expect(testAction).not.toHaveBeenCalled();
            resolve();
          });
        }));

      it('does not dispatch INIT_COMPONENT', () => {
        const actions = store.getActions();
        return expect(actions).toEqual([]);
      });
    });

    describe("with { caller: 'didMount' } option given and init props", () => {
      it('calls the action with the correct values', () => {
        clearComponentIds();
        const mockAction = jest.fn(({ testInitProp }) => Promise.resolve(testInitProp));
        const store = mockStore({ init: { mode: MODE_INIT_SELF, prepared: {} } });
        const MockWithInit = withInitAction(['testInitProp'], { clientOnly: mockAction })(
          SimpleInitTestComponent,
        );

        store.dispatch(
          initComponent(MockWithInit, [45], 'SimpleInitTestComponent[45]', { caller: 'didMount' }),
        );

        return new Promise(resolve => {
          process.nextTick(() => {
            expect(mockAction).toHaveBeenCalledTimes(1);
            expect(mockAction.mock.results[0].value)
              .resolves.toBe(45)
              .then(resolve);
          });
        });
      });

      it('initially dispatches INIT_COMPONENT with { complete: false }', () => {
        clearComponentIds();
        const mockAction = jest.fn(
          ({ testInitProp }) => new Promise(resolve => setTimeout(() => resolve(testInitProp), 40)),
        );
        const store = mockStore({ init: { mode: MODE_INIT_SELF, prepared: {} } });
        const MockWithInit = withInitAction(['testInitProp'], { clientOnly: mockAction })(
          SimpleInitTestComponent,
        );

        store.dispatch(
          initComponent(MockWithInit, [45], 'SimpleInitTestComponent[45]', { caller: 'didMount' }),
        );

        return new Promise(resolve => {
          process.nextTick(() => {
            const actions = store.getActions();
            expect(actions).toMatchSnapshot();

            resolve();
          });
        });
      });

      it('dispatches INIT_COMPONENT with { complete: true } when the action resolves', () => {
        clearComponentIds();
        const mockAction = jest.fn(
          ({ testInitProp }) => new Promise(resolve => setTimeout(() => resolve(testInitProp), 40)),
        );
        const store = mockStore({ init: { mode: MODE_INIT_SELF, prepared: {} } });
        const MockWithInit = withInitAction(['testInitProp'], { clientOnly: mockAction })(
          SimpleInitTestComponent,
        );

        store.dispatch(
          initComponent(MockWithInit, [45], 'SimpleInitTestComponent[45]', { caller: 'didMount' }),
        );

        return new Promise(resolve => {
          process.nextTick(() => {
            mockAction.mock.results[0].value.then(() => {
              process.nextTick(() => {
                const actions = store.getActions();
                expect(actions).toMatchSnapshot();
                resolve();
              });
            });
          });
        });
      });
    });
  });
});
