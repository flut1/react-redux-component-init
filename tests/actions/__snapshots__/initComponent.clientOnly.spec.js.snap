// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`initComponent with clientOnly action when initMode === MODE_INIT_SELF with { caller: 'didMount' } option given and init props dispatches INIT_COMPONENT with { complete: true } when the action resolves 1`] = `
Array [
  Object {
    "payload": Object {
      "complete": false,
      "isPrepare": false,
      "prepareKey": "SimpleInitTestComponent[45]",
    },
    "type": "@@react-redux-component-init/INIT_COMPONENT",
  },
  Object {
    "payload": Object {
      "complete": true,
      "isPrepare": false,
      "prepareKey": "SimpleInitTestComponent[45]",
    },
    "type": "@@react-redux-component-init/INIT_COMPONENT",
  },
]
`;

exports[`initComponent with clientOnly action when initMode === MODE_INIT_SELF with { caller: 'didMount' } option given and init props initially dispatches INIT_COMPONENT with { complete: false } 1`] = `
Array [
  Object {
    "payload": Object {
      "complete": false,
      "isPrepare": false,
      "prepareKey": "SimpleInitTestComponent[45]",
    },
    "type": "@@react-redux-component-init/INIT_COMPONENT",
  },
]
`;
